### PR TITLE
Propagating annotations for fuzzy query [WIP]

### PR DESCRIPTION
**Pull request by alexeyche**
It is a continuation of work regarding fuzzy operator (see https://github.com/vespa-engine/vespa/pull/21689)

This PR introduces two parameters (annotations) to the fuzzy operator:
- `maxEditDistance` which is an inclusive upper bound of an edit distance that informs whether there is a match between the query and the document terms
- `prefixLength` the size of the prefix of a term that is considered non-fuzzy, i.e frozen, it is somewhat similar to the parameter with the same name in the [elasticsearch fuzzy query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-fuzzy-query.html#fuzzy-query-field-params)

In `system-test` repo I put a PR to more thoroughly test the logic of how these parameters should behave:
https://github.com/vespa-engine/system-test/pull/2272

In order to make prefix matching work for other cases but `fast-search` I was needed manually implement prefix match within the `FuzzyMatcher`, so the logic there was altered a bit.

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Review by geirst — COMMENTED** [review]
This looks very good. @bratseth Can you have a look at the Java code?

**Review by bratseth — APPROVED** [review]
LGTM!

File: searchlib/src/vespa/searchlib/query/streaming/queryterm.h [hunk 1]
```diff
@@ -85,6 +85,10 @@ class QueryTerm : public QueryTermUCS4, public QueryNode
     void visitMembers(vespalib::ObjectVisitor &visitor) const override;
     void setIndex(const string & index_) override { _index = index_; }
     const string & getIndex() const override { return _index; }
+    void setFuzzyMaxEditDistance(uint32_t fuzzyMaxEditDistance) { _fuzzyMaxEditDistance = fuzzyMaxEditDistance; }
+    void setFuzzyPrefixLength(uint32_t fuzzyPrefixLength) { _fuzzyPrefixLength = fuzzyPrefixLength; }
+    uint32_t fuzzyMaxEditDistance() const { return _fuzzyMaxEditDistance; }
```
[the remark this answers is not in the document]
Comment (alexeyche) [reply searchlib/src/vespa/searchlib/query/streaming/queryterm.h:90]: It's not needed, will remove it

**Comment by alexeyche** [comment]
@bratseth @geirst Addressed your comments in a separate commit. Thanks for the review!

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**